### PR TITLE
[ENG-158] feat: Abuse protection for OTP based login

### PR DESCRIPTION
**Pull request by praffq**
## Proposed Changes

- enchancements to enforce more strict control over OTP

### Associated Issue

- Link to issue here, explain how the proposed solution will solve the reported issue/ feature request.

### Architecture changes

- Remove this section if not used

## Merge Checklist

- [ ] Tests added/fixed
- [ ] Update docs in `/docs`
- [ ] Linting Complete
- [ ] Any other necessary step

_*Only PR's with test cases included and passing lint and test pipelines will be reviewed*_

@ohcnetwork/care-backend-maintainers @ohcnetwork/care-backend-admins

## Summary by CodeRabbit

* **New Features**
  * Per-phone resend limits, per-OTP verification caps, and atomic validation of the latest unused OTP
  * Successful OTP login now returns an access token and consumes the OTP

* **Chores**
  * Configurable OTP policies (send window, send cap, verify attempts, failures, lockout, validity) and DB tracking for failed attempts
  * Periodic cleanup to expire old OTPs; example env vars updated

* **Tests**
  * End-to-end tests covering send/login flows, rate limits, lockout, retries, and edge cases

### CODE REVIEW

**Review by sainak — APPROVED** [review]

File: care/emr/locks/otp.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+from django.conf import settings
+
+from care.utils.lock import Lock
+
+
+class OTPSendLock(Lock):
```
[the remark this answers is not in the document]
Comment (praffq) [reply care/emr/locks/otp.py:6]: > This class does not call [Lock.init](https://github.com/ohcnetwork/care/blob/c8eb2e32e13aa56942e100c9be7736b7b9cfa534/care/utils/lock.py#L13-L13) during initialization. ([OTPSendLock.init](https://github.com/ohcnetwork/care/blob/c8eb2e32e13aa56942e100c9be7736b7b9cfa534/care/emr/locks/otp.py#L7-L7) may be missing a call to a base class init) 

This is a valid and expected, we are overwriting the Base class init to modify the keys according to the subclass

File: care/emr/api/otp_viewsets/login.py [hunk 2]
```diff
@@ -47,42 +50,64 @@ class OTPLoginView(EMRBaseViewSet):
...
+                except Exception as e:
+                    logger.error(e)
+                    return Response(
+                        {"error": "Error while sending OTP. Contact admin."},
+                        status=400,
+                    )
+            elif settings.IS_PRODUCTION:
+                random_otp = rand_pass(settings.OTP_LENGTH)
+            else:
+                random_otp = "45612"
```
Comment (sainak) [inline care/emr/api/otp_viewsets/login.py]:
```suggestion
            elif not settings.IS_PRODUCTION:
                random_otp = "45612"
            else:
                raise APIException("SMS Backend not configured")
```
No point in generating OTP for deployments without SMS service

File: care/emr/api/otp_viewsets/login.py [hunk 3]
```diff
@@ -47,42 +50,64 @@ class OTPLoginView(EMRBaseViewSet):
...
+                    timezone.now() - timedelta(minutes=settings.OTP_SEND_WINDOW_MINUTES)
+                ),
+                phone_number=data.phone_number,
+            )
+            if sent_otps.count() >= settings.OTP_MAX_SENDS_PER_WINDOW:
+                raise ValidationError({"phone_number": "Max Retries has exceeded"})
+
+            random_otp = ""
+            if settings.USE_SMS:
+                random_otp = rand_pass(settings.OTP_LENGTH)
```
Comment (sainak) [inline care/emr/api/otp_viewsets/login.py]:
```suggestion
                random_otp = rand_numbers(settings.OTP_LENGTH)
```
We can rename this function

File: care/emr/api/otp_viewsets/login.py [hunk 4]
```diff
@@ -47,42 +50,64 @@ class OTPLoginView(EMRBaseViewSet):
...
     permission_classes = []
 
+    def failure_count(self, phone_number: str) -> int:
+        since = timezone.now() - timedelta(minutes=settings.OTP_LOCKOUT_MINUTES)
+        total = PatientMobileOTP.objects.filter(
+            phone_number=phone_number,
+            modified_date__gte=since,
+            failed_attempts__gt=0,
+        ).aggregate(total=Sum("failed_attempts"))["total"]
+        return total or 0
```
Comment (sainak) [inline care/emr/api/otp_viewsets/login.py]: Is this query actually using pmo_phone_modified_idx? Did you run analyze on the generated query?

File: care/emr/api/otp_viewsets/login.py [hunk 5]
```diff
@@ -91,16 +116,36 @@ def send(self, request):
...
+                .first()
+            )
+
+            if otp_object:
+                if otp_object.otp == data.otp:
+                    otp_object.is_used = True
+                    otp_object.save(update_fields=["is_used", "modified_date"])
+                    token = PatientToken()
+                    token["phone_number"] = data.phone_number
+                    return Response({"access": str(token)})
```
Comment (sainak) [inline care/emr/api/otp_viewsets/login.py:166]: We should check if the otp was generated long ago

Comment (sainak) [inline care/emr/api/otp_viewsets/login.py:166]: We should invalidate the other otps returned by the otp_object query

File: config/settings/base.py [hunk 6]
```diff
@@ -502,10 +502,18 @@
...
+OTP_SEND_WINDOW_MINUTES = env.int("OTP_SEND_WINDOW_MINUTES", default=60)
+# max no. of otp that can be generated in (OTP_SEND_WINDOW_MINUTES)mins
+OTP_MAX_SENDS_PER_WINDOW = env.int("OTP_MAX_SENDS_PER_WINDOW", default=10)
+# no. of attempts after which the otp is blocked
+OTP_MAX_VERIFY_ATTEMPTS = env.int("OTP_MAX_VERIFY_ATTEMPTS", default=3)
+# max no. of attempts after which the phone number is blocked
+OTP_MAX_FAILURES = env.int("OTP_MAX_FAILURES", default=5)
+# minutes it stays lock when reached OTP_MAX_FAILURES
+OTP_LOCKOUT_MINUTES = env.int("OTP_LOCKOUT_MINUTES", default=30)
+
```
Comment (sainak) [inline config/settings/base.py:524]:
```suggestion
# The time window (in minutes) for tracking OTP request limits
OTP_SEND_WINDOW_MINUTES = env.int("OTP_SEND_WINDOW_MINUTES", default=60)

# Maximum number of OTPs that can be generated within the window defined above
OTP_MAX_SENDS_PER_WINDOW = env.int("OTP_MAX_SENDS_PER_WINDOW", default=10)

# Number of failed verification attempts allowed before an OTP is invalidated
OTP_MAX_VERIFY_ATTEMPTS = env.int("OTP_MAX_VERIFY_ATTEMPTS", default=3)

# Maximum total failures allowed before the phone number is restricted
OTP_MAX_FAILURES = env.int("OTP_MAX_FAILURES", default=5)

# Duration (in minutes) the account remains locked after reaching maximum failures
OTP_LOCKOUT_MINUTES = env.int("OTP_LOCKOUT_MINUTES", default=60)

```

Updated comments and otp lockout window
do add these to .env.example

File: care/facility/models/patient.py [hunk 7]
```diff
@@ -10,3 +10,16 @@ class PatientMobileOTP(BaseModel):
...
+
+    class Meta:
+        indexes = [
+            models.Index(
+                fields=["phone_number", "-created_date"],
+                name="pmo_phone_created_idx",
+            ),
+            models.Index(
+                fields=["phone_number", "-modified_date"],
+                name="pmo_phone_modified_idx",
```
Comment (sainak) [inline care/facility/models/patient.py]: Make sure that the indexes are actually being used
we are not clearing OTPs, so these will keep growing

Another approach is to add an expired column, and use that to create a partial index

File: care/emr/api/otp_viewsets/login.py (hunk 4 above)
Comment (praffq) [reply care/emr/api/otp_viewsets/login.py]: checked using debugger (to get sql used) and EXPLAIN ANALYSE , both does uses the index in querying

1.  
```
care=# EXPLAIN ANALYZE
care-# SELECT
care-#     SUM("facility_patientmobileotp"."failed_attempts") AS "total"
care-# FROM
care-#     "facility_patientmobileotp"
care-# WHERE
care-#     NOT "facility_patientmobileotp"."deleted"
care-#     AND "facility_patientmobileotp"."failed_attempts" > 0
care-#     AND "facility_patientmobileotp"."modified_date" >= '2026-05-06 16:14:21.732414+00:00'::timestamptz
care-#     AND "facility_patientmobileotp"."phone_number" = '+919342962310';
                                                                       QUERY PLAN                                                                       
--------------------------------------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=8.17..8.18 rows=1 width=8) (actual time=0.040..0.046 rows=1 loops=1)
   ->  Index Scan using pmo_phone_modified_idx on facility_patientmobileotp  (cost=0.15..8.17 rows=1 width=2) (actual time=0.039..0.039 rows=0 loops=1)
         Index Cond: (((phone_number)::text = '+919342962310'::text) AND (modified_date >= '2026-05-06 16:14:21.732414+00'::timestamp with time zone))
         Filter: ((NOT deleted) AND (failed_attempts > 0))
         Rows Removed by Filter: 1
 Planning Time: 1.403 ms
 Execution Time: 0.152 ms
(7 rows)
```

2. 
```
care=# 
care=# EXPLAIN ANALYZE
care-# SELECT
care-#     COUNT(*) AS "__count"
care-# FROM
care-#     "facility_patientmobileotp"
care-# WHERE
care-#     NOT "facility_patientmobileotp"."deleted"
care-#     AND "facility_patientmobileotp"."created_date" >= '2026-05-06 16:14:21.748974+00:00'::timestamptz
care-#     AND "facility_patientmobileotp"."phone_number" = '+919342962310';
                                                                      QUERY PLAN                                                                       
-------------------------------------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=8.17..8.18 rows=1 width=8) (actual time=0.087..0.089 rows=1 loops=1)
   ->  Index Scan using pmo_phone_created_idx on facility_patientmobileotp  (cost=0.15..8.17 rows=1 width=0) (actual time=0.079..0.081 rows=1 loops=1)
         Index Cond: (((phone_number)::text = '+919342962310'::text) AND (created_date >= '2026-05-06 16:14:21.748974+00'::timestamp with time zone))
         Filter: (NOT deleted)
 Planning Time: 0.393 ms
 Execution Time: 0.138 ms
(6 rows)
```

File: care/emr/api/otp_viewsets/login.py (hunk 5 above)
Comment (praffq) [reply care/emr/api/otp_viewsets/login.py:166]: we won't be having multiple active otps at a time at the time of login. We invalidate all the old otps there

File: care/facility/models/patient.py (hunk 7 above)
Comment (praffq) [reply care/facility/models/patient.py]: Added partial index with deleted=False

Also added a cron job to soft delete the stable otp rows (similar to what we do for token but doing soft delete since otp record maybe require to know past login activity of patient). This ensures the index doesn't grow much

**Review by sainak — APPROVED** [review]

File: care/emr/tasks/cleanup_expired_otps.py [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+logger: Logger = get_task_logger(__name__)
+
+
+@shared_task
+def cleanup_expired_otps():
+    """
+    Soft-deletes MobileOTP rows older than the lockout window
+    """
+    cutoff = care_now() - timedelta(minutes=settings.OTP_LOCKOUT_MINUTES)
+    count = MobileOTP.objects.filter(created_date__lt=cutoff).update(deleted=True)
```
Comment (vigneshhari) [inline care/emr/tasks/cleanup_expired_otps.py]: Lets hard delete, no point in keeping these

File: care/emr/api/otp_viewsets/login.py [hunk 9]
```diff
@@ -105,16 +154,41 @@ def send(self, request):
...
 
-        token = PatientToken()
-        token["phone_number"] = data.phone_number
+        if self.failure_count(data.phone_number) >= settings.OTP_MAX_FAILURES:
+            raise Throttled(detail="Too many failed login attempts. Try again later.")
+
+        expired = False
+        with transaction.atomic():
+            otp_object = (
+                MobileOTP.objects.select_for_update()
```
Comment (vigneshhari) [inline care/emr/api/otp_viewsets/login.py]: Lets use locks in redis instead of select_for_update

File: care/emr/api/otp_viewsets/login.py [hunk 10]
```diff
@@ -85,18 +92,60 @@ class OTPLoginView(EMRBaseViewSet):
...
-        except ValueError as e:
-            raise ValidationError({"phone_number": "Unable to send OTP"}) from e
-        except Exception:
-            return Response({"error": "Unable to send OTP"}, status=400)
+
+        if self.failure_count(data.phone_number) >= settings.OTP_MAX_FAILURES:
+            raise Throttled(detail="Too many failed login attempts. Try again later.")
+
+        with OTPSendLock(data.phone_number):
+            sent_otps = MobileOTP.objects.filter(
```
Comment (nandkishorr) [inline care/emr/api/otp_viewsets/login.py]: Can't we can reuse the `def send_otp()` here right ?

**Review by vigneshhari — CHANGES_REQUESTED** [review]

File: care/users/api/otp_viewset/reset_password.py [hunk 11]
```diff
@@ -46,12 +54,7 @@ def send(self, request):
         if not User.objects.filter(phone_number=data.phone_number).exists():
             return Response({"otp": "generated"})
 
-        try:
-            send_otp(data.phone_number, otp_type=ResetPasswordOTP)
```
Comment (vigneshhari) [inline care/users/api/otp_viewset/reset_password.py:50]: Why is the validations removed ?

**Review by vigneshhari — CHANGES_REQUESTED** [review]

File: care/users/api/otp_viewset/reset_password.py (hunk 11 above)
Comment (praffq) [reply care/users/api/otp_viewset/reset_password.py:50]: Validation are moved into send_otp

File: care/emr/locks/otp.py [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+    def __init__(self, phone_number, timeout=settings.LOCK_TIMEOUT):
+        self.key = f"lock:otp_send:{phone_number}"
+        self.timeout = timeout
+
+
+class OTPVerifyLock(Lock):
```
[the remark this answers is not in the document]
Comment (praffq) [reply care/emr/locks/otp.py:12]: This is a valid and expected, we are overwriting the Base class init to modify the keys according to the subclass

File: care/emr/api/otp_viewsets/login.py [hunk 13]
```diff
@@ -81,6 +72,54 @@ class OTPLoginSpec(OTPRequestBaseSpec):
...
+            created_date__gte=care_now() - otp_type.send_window(),
+            phone_number=phone_number,
+        )
+        if sent_otps.count() >= otp_type.max_sends():
+            raise ValidationError({"phone_number": "Max Retries has exceeded"})
+
+        random_otp = ""
+        if settings.USE_SMS:
+            random_otp = generate_otp(settings.OTP_LENGTH)
+            try:
```
Comment (nandkishorr) [inline care/emr/api/otp_viewsets/login.py:101]: Use USE_SMS only to control whether the SMS sending function is invoked, and keep IS_PRODUCTION focused on environment-specific, the OTP should be randomly generated or fixed.

**Review by nandkishorr — CHANGES_REQUESTED** [review]

**Comment by vigneshhari** [comment]
Can you confirm that the reset password flow also has abuse protection ?

**Review by nandkishorr — APPROVED** [review]

**Review by vigneshhari — APPROVED** [review]
